plugins/provisioners/puppet/config/puppet.rb

```diff
@@ -95,10 +102,15 @@ def finalize!
             @environment_variables = {}
           end
 
+          if @synced_folder_opts.eql?({})
+            @synced_folder_opts[:type] = @synced_folder_type if @synced_folder_type
+            @synced_folder_opts[:owner] = "root" if !@synced_folder_type
+            @synced_folder_opts[:args] = @synced_folder_args if @synced_folder_args
+            @synced_folder_opts[:nfs__quiet] = true
+          end
```

Comment:
Woot! This is a general comment not for here but for the `validate` method.

I think we should print a warning that checks if both `:type` and `@synced_folder_type` are set just so it's clear what is getting overridden here in case both options are set. If they're different, we should probably just make that an error?

Reply:
Hm, unsure, because there is no need to set ```:type``` within the ```synced_folder_opts``` hash. This is intended to be exactly the same hash symbols that are supported by any ```synced_folder``` type, like e.g.
```
synced_folder_opts :nfs => true, :nfs_udp => false, :nfs_version => 4, :nfs_export => false
```
which will be translated to ```:type``` within the ```vm.synced_folder``` method i guess and then if someone explicitly sets ```synced_folder_opts``` the one would probably know that there is no need to mix up with the deprecated and then redundant ```synced_folder_type``` option. But of course if you insist i will implement it like you wish.

And then maybe should i include the example within the website documentation, so to make it clear that there is no need to set both of the options?

Reply:
So these options get directly passed into the `synced_folder` config method: https://github.com/hashicorp/vagrant/blob/master/plugins/kernel_v2/config/vm.rb#L204

Which means you should be able to give it any of the options that are valid here: https://www.vagrantup.com/docs/synced-folders/basic_usage.html#options

So there's a possibility that someone could set `{type: "nfs"}` as a key like

```ruby
config.vm.provider "puppet" do |p|
  p.synced_folder_type = "nfs"
  p.synced_folder_opts = {type: "smb"}
end
```

Which is silly of course, but I think it would be good to warn users if both happen to be set is all.

But yeah in the docs it should specifically call out the doc page I link to as all the valid options you can provide to `synced_folder_opts`.

Reply:
Ok, i was too lazy to code for silly folks ;) but i will follow your advice here as well.
I did already link to the ```basic_usage.html``` page though in the documentation but will add the ```#options``` to it. 
Do you think i should add an extra section within the example section of the provisioner documentation page?

Reply:
AH you're right I'm not sure if that's what we want then. I could see users expecting `synced_folder_type` to work as well as `synced_folder_opts` with other various options. Like maybe it's an old Vagrantfile with `synced_folder_type` set, and then they wanted to add new options like `mount_options` later and realize their `synced_folder_type` param will be ignored.

Reply:
I also do think it's probably worth while to give a simple example of using the new `synced_folder_opts` option for the puppet provisioner.

Reply:
Ok, will add an example to the documentation then.

As for the merging of ```synced_folder_type``` i'm still not fully convinced, if someone is able to find the new ```synced_folder_opts``` option and set the values, they will probably have been reading that the older ```synced_folder_type``` is deprecated. 
If we really check against collisions we'll have to check against all the synced_folder specific symbols as well like ```:nfs => true``` or the like, that is a lot of code to support an option that is deprecated in favor of a more capable option (and currently i don't understand from the code what will happen in the case when both ```:nfs => true, :type => "rsync"``` are given).

I agree that a warning should be given when both options are present though. What do you think?

Reply:
So the `:nfs` key for the puppet provisioner config has been deprecated for a while now, and it's no longer even on the puppet provisioner docs. And `:nfs=>true` is not a valid `synced_folder` option. So we should be ok to just check for `:type` in `synced_folder_opts` and `synced_folder_type` existing instead. Anyone still using `:nfs=>true` on the puppet provisioner has seen that warning for a while now and shouldn't be too surprised if it is gone on an upgrade.

But yeah, ultimately we should display a warning if both `type` and `synced_folder_type` is present, and it should tell the user that the `synced_folder_opts` `:type` key will actually be used and that `synced_folder_type` will be ignored.

> (and currently i don't understand from the code what will happen in the case when both :nfs => true, :type => "rsync" are given)

Yep...unfortunately if both are set, it is dependent on what  _order_  they are defined in the Vagrantfile. Because `#nfs=` also sets `synced_folder_type`, which ever option is defined _last_ in the provider config block will be the one that wins out setting that variable. Not a great user experience here, but I think your PR here will make this better!

Reply:
Sorry for the confusion, should have made clear which keys/options i mean. I was always talking about the keys given to ```synced_folder_opts``` here if not referenced with ```synced_folder_type```. 

So i guess i'm going to respect and merge the ```synced_folder_type``` as ```:type => @synced_folder_type``` into ```synced_folder_opts``` if none of the keys ```:nfs | :rsync | :smb | :virtualbox | :type``` are present within ```synced_folder_opts``` and discard ```synced_folder_type``` if the opposite is true.

Same goes for the ```synced_folder_args``` but only if ```:type``` will result in ```rsync```, because i think this option is only applicable to rsync folders (https://www.vagrantup.com/docs/provisioning/puppet_apply.html#synced_folder_args)

I will give a warning if collisions occur, any of the two deprecated options are set, or when no ```:type``` is given at all.

Will need some time right now to setup my rubymine environment properly now because that is not playing nice with NixOs as it is, but will for sure finish this PR properly.

Reply:
Sounds great, thank you! Let me know if you need any more help finishing it :tada: